File: rust/src/ArrayList.rs
use std::collections::VecDeque;

#[derive(Debug)]
pub struct ArrayList<T> {
    length: usize,
    items: VecDeque<T>,
}

impl<T> ArrayList<T> {
    pub fn new() -> Self {
        Self {
            length: 0,
            items: VecDeque::new(),
        }
    }

    pub fn prepend(&mut self, item: T) {
        self.items.push_front(item);
        self.length += 1;
    }

    pub fn insert_at(&mut self, item: T, idx: usize) {
        if idx < 0 || idx > self.length {
            panic!("Index out of bounds");
        }

        self.items.insert(idx, item);
        self.length += 1;
    }

    pub fn append(&mut self, item: T) {
        self.items.push_back(item);
        self.length += 1;
    }

    pub fn remove(&mut self, item: T) -> Option<T> {
        let idx = self.items.iter().position(|x| x == item);
        if let None = idx {
            return None;
        }

        self.items.remove(idx.unwrap());
        self.length -= 1;
        Some(item)
    }

    pub fn get(&self, idx: usize) -> Option<&T> {
        if idx < 0 || idx >= self.length {
            return None;
        }

        self.items.get(idx)
    }

    pub fn remove_at(&mut self, idx: usize) -> Option<T> {
        if idx < 0 || idx >= self.length {
            return None;
        }

        let item = self.items.remove(idx);
        self.length -= 1;
        Some(item)
    }
}

File: java/src/ArrayList.java
package src;

import java.lang.reflect.Array;

public class ArrayList<T> {
    private int length;
    private T[] items;

    public ArrayList(Class<T> clazz) {
        this.length = 0;
        this.items = (T[]) Array.newInstance(clazz, 0);
    }

    void prepend(T item) {
        T[] newItems = (T[]) Array.newInstance(item.getClass(), length + 1);
        System.arraycopy(items, 0, newItems, 1, length);
        newItems[0] = item;
        items = newItems;
        length++;
    }

    void insertAt(T item, int idx) {
        if (idx < 0 || idx > this.length) {
            throw new RuntimeException("Index out of bounds");
        }

        T[] newItems = (T[]) Array.newInstance(item.getClass(), length + 1);
        System.arraycopy(items, 0, newItems, 0, idx);
        System.arraycopy(items, idx, newItems, idx + 1, length - idx);
        newItems[idx] = item;
        items = newItems;
        length++;
    }

    void append(T item) {
        T[] newItems = (T[]) Array.newInstance(item.getClass(), length + 1);
        System.arraycopy(items, 0, newItems, 0, length);
        newItems[length] = item;
        items = newItems;
        length++;
    }

    T remove(T item) {
        int idx = indexOf(item);
        if (idx == -1) {
            return null;
        }

        T[] newItems = (T[]) Array.newInstance(item.getClass(), length - 1);
        System.arraycopy(items, 0, newItems, 0, idx);
        System.arraycopy(items, idx + 1, newItems, idx, length - idx - 1);
        items = newItems;
        length--;
        return item;
    }

    T get(int idx) {
        if (idx < 0 || idx >= this.length) {
            return null;
        }

        return this.items[idx];
    }

    T removeAt(int idx) {
        if (idx < 0 || idx >= this.length) {
            return null;
        }

        T removedItem = items[idx];
        T[] newItems = (T[]) Array.newInstance(removedItem.getClass(), length - 1);
        System.arraycopy(items, 0, newItems, 0, idx);
        System.arraycopy(items, idx + 1, newItems, idx, length - idx - 1);
        items = newItems;
        length--;
        return removedItem;
    }

    private int indexOf(T item) {
        for (int i = 0; i < length; i++) {
            if (items[i].equals(item)) {
                return i;
            }
        }
        return -1;
    }
}
